### ISSUE TITLE
`unused_io_amount` false positive if the result is matched by `Ok(<literal>)` or `unreachable!()` cases

### ISSUE THREAD

**Issue by oxalica** [rust-lang/rust-clippy#12208]
### Summary

See the reproducer. I believe it's introduced by #12005. cc @PartiallyTyped

I saw there are two issues:
1. We should consider `Ok(<literal>)` as used, as the read amount is actually checked.
2. We should consider `Ok(_) => unreachable!()` as used, as we are explicitly asserting the read must fail for some reason. I'm not quite sure how exactly we should do this. As simply checking if the arm has type `!` would instead yield a false-negative for `Ok(_) => return`. 

### Lint Name

unused_io_amount

### Reproducer

Given this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fb571019516a72fb6b51e9c0ea40ae14)):
```rust
pub fn work(rdr: &mut dyn std::io::Read) {
    match rdr.read(&mut [0]) {
        Ok(0) => println!("EOF"),
        Ok(_) => println!("fully read"),
        Err(_) => println!("fail"),
    }
}

pub fn work2(rdr: &mut dyn std::io::Read) {
    match rdr.read(&mut [0]) {
        Ok(_) => unreachable!(),
        Err(_) => println!("expected"),
    }
}
```

I saw this happen:

```
error: read amount is not handled
 --> src/lib.rs:2:11
  |
2 |     match rdr.read(&mut [0]) {
  |           ^^^^^^^^^^^^^^^^^^
  |
  = help: use `Read::read_exact` instead, or handle partial reads
note: the result is consumed here, but the amount of I/O bytes remains unhandled
 --> src/lib.rs:4:9
  |
4 |         Ok(_) => println!("fully read"),
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_io_amount
  = note: `#[deny(clippy::unused_io_amount)]` on by default

error: read amount is not handled
  --> src/lib.rs:10:11
   |
10 |     match rdr.read(&mut [0]) {
   |           ^^^^^^^^^^^^^^^^^^
   |
   = help: use `Read::read_exact` instead, or handle partial reads
note: the result is consumed here, but the amount of I/O bytes remains unhandled
  --> src/lib.rs:11:9
   |
11 |         Ok(_) => unreachable!(),
   |         ^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_io_amount
```

I expected to see this happen: no errors or warnings emitted.

### Version

```text
(Playground) 0.1.77 (2024-01-27 6b4f1c5)
```

**Comment by PartiallyTyped** [comment]
@blyxyas what do you think?

I am not really sure what the "right" way to handle this is tbh; and I think I just opened a whole new can of worms 🤔.

**Comment by FreezyLemon** [comment]
Hey, not sure if this is the exact same problem but it seems related. Calling `.read` inside a block expression also triggers a false positive ([Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3c23747b00784881bdf9031be6aec92a)):

```rust
pub fn work(rdr: &mut dyn std::io::Read) -> std::io::Result<usize> {
    let read = { rdr.read(&mut [0])? };
    Ok(read)
}
```

This does not happen on stable. Removing the braces will fix it:

```rust
pub fn work2(rdr: &mut dyn std::io::Read) -> std::io::Result<usize> {
    let read = rdr.read(&mut [0])?;
    Ok(read)
}
```

**Comment by blyxyas** [comment]
> what do you think?

The reporter is right, `Ok(<literal>)` should not lint, that should be easy to implement. I'm not sure how to handle `Ok(_) => <panic>` thought. False negatives with `//TODO`s are better than false positives in almost all cases.

I've put this issue in Priority High, we should address it pretty quickly (before it reaches stable)

**Comment by PartiallyTyped** [comment]
@blyxyas 

Ok, so in case any literal exists on an ok arm we will consider this covered.

What about cases with arm guards?

**Comment by blyxyas** [comment]
I'd say that an arm guard would mean that it's covered. Maybe the guard calls a function to do the checking (unnecessarily complex, but it's still handling)

**Comment by m0rphej** [comment]
This should fix it. Thanks for the reports, and to any future readers, please do tag me and sorry for any troubles.

**Comment by oxalica** [comment]
It seems that this bug goes into stable 1.77.0, but the fix does not. Thus the code in issue description generates warnings on stable but not on nightly. Should we reopen this?

**Comment by m-rph** [comment]
Seems like CR wasnt labelled with beta nominated. 

@flip1995 any ideas?

**Comment by flip1995** [comment]
1.77.0 was already released, so there's no way to fix this in stable now. In the unlikely event of a point release, I try to include this fix there, but can't promise that it will be accepted. But a Clippy bug fix alone, is not a reason for a point release.
